### PR TITLE
Use git-describe in versions

### LINKED ISSUE #291: Use commit hashes in versions
- [x] I have read the specified guidelines for issues

**Is your feature request related to a problem? Please describe.**

Tenacity recently removed dates from versions in order to make builds more reproducible. See fbaa87b5110ef2029b10f6d4525d13cf24a0c49d, but the version is very vague now.

**Describe the solution you'd like**
It was discussed that we should use commit hashes instead.

**Additional context**
See https://github.com/tenacityteam/tenacity/pull/224

### PR DESCRIPTION
**Pull request by nbsp**
Resolves: #291
Closes: #413 

Use `git describe --abbrev=7` in versions.
This also eliminates the need for `${AUDACITY_VERSION}.${AUDACITY_RELEASE}.${AUDACITY_REVISION}`, as git-describe comes with the latest tag already included.

```
# Before
X.Y.Z-alpha-20210615+a1b2c3d

# After
X.Y.Z-alpha-1337-ga1b2c3d
```

I still need to clean up some artifacts of the previous versioning system.

<details>
<summary>Checklist</summary>

- [x] I have signed off my commits using `-s` or `Signed-off-by`\* (See: [Contributing § DCO](https://github.com/tenacityteam/tenacity/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code\*
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving\*
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"\*

\* indicates required

</details>

### CODE REVIEW

**Comment by nbsp** [comment]
Still likely unfinished, but I'd like to have someone review it.

**Comment by Semisol** [comment]
why abbrev=5? I think that is pretty low, like 120k commits max, and collisions can happen (see the birthday paradox) so can't we have abbr 8-12 at least?

**Comment by nbsp** [comment]
> why abbrev=5? I think that is pretty low, like 120k commits max, and collisions can happen (see the birthday paradox) so can't we have abbr 8-12 at least?

The other parts of `git describe` make sure there isn't one. It's the latest annotated tag, followed by the number of commits after it, and then the short commit hash. I don't see us rewriting history to the same commit 120k times, and `abbrev=5` is the shortest I could make this hash.

**Comment by Semisol** [comment]
Hmm, okay then

**Comment by AnotherFoxGuy** [comment]
How would you find the commit it was built with in this PR? it doesn't seem to have the commit info anymore

**Comment by nbsp** [comment]
> How would you find the commit it was built with in this PR? it doesn't seem to have the commit info anymore

I'm not sure I understand the question. The `git describe` output might differ between PR and production, because rebases and merges might change the commit hash or the number of commits since a tag, but we're not comparing the two. The important part is that a prebuilt artifact of a certain commit will always have the exact same `git describe` output as when you build it yourself.

**Comment by AnotherFoxGuy** [comment]
Before this PR the version format was `X.Y.Z-alpha-20210615+ca2977039` where ca2977039 was the commit that was built
And with this PR the version format is `X.Y.Z-1743-gff947-alpha`, but gff947 isnt a commit

**Comment by Semisol** [comment]
> Before this PR the version format was `X.Y.Z-alpha-20210615+ca2977039` where [ca29770](https://github.com/tenacityteam/tenacity/commit/ca2977039b671cac5c7afc43bdb30984c62442de) was the commit that was built
> And with this PR the version format is `X.Y.Z-1743-gff947-alpha`, but gff947 isnt a commit

The example didn't comply to hex

**Comment by nbsp** [comment]
That's not the case: if you'd looked in the man page or played around with `git describe`, you'd have seen that "g" is a prefix for all of the hashes.

... and this is commit `ff947`.

**Comment by AnotherFoxGuy** [comment]
I never used describe in my projects, I usually use `git rev-parse --short HEAD`

File: CMakeLists.txt [hunk 1]
```diff
@@ -233,14 +233,12 @@ elseif( CMAKE_SYSTEM_NAME MATCHES "Darwin" )
...
-set( GIT_COMMIT_SHORT "unknown" )
-set( GIT_COMMIT_LONG "unknown" )
-set( GIT_COMMIT_TIME "unknown" )
+set( GIT_DESCRIBE "unknown" )
 find_package( Git QUIET )
 if( GIT_FOUND )
    execute_process(
       COMMAND
-         ${GIT_EXECUTABLE} show -s "--format=%h;%H;%cd" --no-show-signature --no-notes
+         ${GIT_EXECUTABLE} describe --abbrev=5
```
Comment (Be-ing) [inline CMakeLists.txt]:
```suggestion
         ${GIT_EXECUTABLE} describe --abbrev=7
```
How about using the same as the default Git short hash length?

File: CMakeLists.txt [hunk 2]
```diff
@@ -250,12 +248,10 @@ if( GIT_FOUND )
    )
 
    if( git_output )
-      list( GET git_output 0 GIT_COMMIT_SHORT )
-      list( GET git_output 1 GIT_COMMIT_LONG )
-      list( GET git_output 2 GIT_COMMIT_TIME )
+       list( GET git_output 0 GIT_DESCRIBE )
```
Comment (Be-ing) [inline CMakeLists.txt]: There is no need for this to be a list anymore.

**Comment by Be-ing** [comment]
> X.Y.Z-1743-gff947-alpha

This is ugly because it defeats part of the advantage of using `git describe`. The output of `git describe` can be copied and pasted into any Git command that takes a commit reference, for example `git log` or `git show`. In Mixxx we are making `-alpha` and `-beta` tags so that is included in the `git describe` output. We could do that for Tenacity, or alternatively we could have the "alpha"/"beta" string shown separately from the `git describe` string.

File: CMakeLists.txt (hunk 1 above)
Comment (nbsp) [reply CMakeLists.txt]: Sure.

**Comment by nbsp** [comment]
> > X.Y.Z-1743-gff947-alpha
> 
> This is ugly because it defeats part of the advantage of using `git describe`. The output of `git describe` can be copied and pasted into any Git command that takes a commit reference, for example `git log` or `git show`. In Mixxx we are making `-alpha` and `-beta` tags so that is included in the `git describe` output. We could do that for Tenacity, or alternatively we could have the "alpha"/"beta" string shown separately from the `git describe` string.

I don't really understand why the `-alpha`/`-beta` distinction is even there, and I agree that it should be handled by git entirely, but I'm not sure if it'd break anything.

**Comment by Be-ing** [comment]
> I'm not sure if it'd break anything

Let's try getting rid of it and see if it breaks anything.

**Comment by nbsp** [comment]
I think it works fine, but I'd like to have it tested it on Windows or macOS.

**Comment by Semisol** [comment]
> I think it works fine, but I'd like to have it tested it on Windows or macOS.

on it ;)

**Comment by nbsp** [comment]
Still haven't gotten rid of `AUDACITY_{VERSION|RELEASE|REVISION|MODLEVEL}`, because they're very tightly integrated into the build process. Might want to look at `git describe --abbrev=0` for that.

**Review by Semisol — DISMISSED** [review]
Works on my machine.

- [x] macOS
- [ ] Linux
	- [x] Arch
	- [ ] Ubuntu
	- [ ] Fedora
	- [ ] Alpine
	- [ ] Something else (add it here!)
- [x] Windows

**Comment by nbsp** [comment]
Works on my Windows 10 VM.

**Comment by nbsp** [comment]
I haven't managed to delete them entirely, but I have removed the manual setting of `AUDACITY_{VERSION|RELEASE|REVISION}`. I think this is good enough for now.

**Comment by nbsp** [comment]
I'm... not sure why this doesn't work? It shows no errors on my machine.
It also works just fine on the Arch build, what

File: CMakeLists.txt [hunk 3]
```diff
@@ -18,25 +18,45 @@ endif ()
...
-else()
-   set( AUDACITY_SUFFIX "" )
+set( AUDACITY_REVISION 0 )
+find_package( Git QUIET )
+if( GIT_FOUND )
+    execute_process(
+        COMMAND
+            ${GIT_EXECUTABLE} describe --abbrev=0
+        WORKING_DIRECTORY
+            ${topdir}
```
Comment (Be-ing) [inline CMakeLists.txt]:
```suggestion
            ${CMAKE_SOURCE_DIR}
```
This `${topdir}` variable is overcomplicated. It does not make the code easier to read IMO; it only makes me wonder why the `${topdir}` variable exists and look up where it is set to understand what it does.

Comment (nbsp) [reply CMakeLists.txt]: `libsrc` isn't used very much either, but I think that should be a separate cleanup PR.

Comment (Be-ing) [reply CMakeLists.txt]: I agree. Here I am only asking not to use it in new code.

Comment (nbsp) [reply CMakeLists.txt]: Fair. I'll put it back where it was then

Comment (Be-ing) [reply CMakeLists.txt]: You put the old code back where it was but the new code is still using `${topdir}`

Comment (nbsp) [reply CMakeLists.txt]: Yes, I fixed only one instance of it. one sec

**Review by Be-ing — DISMISSED** [review]

**Comment by nbsp** [comment]
This revision should delete `-alpha`/`-beta` from git tags as well.

File: CMakeLists.txt [hunk 4]
```diff
@@ -22,21 +22,38 @@ endif ()
...
-elseif(  AUDACITY_BUILD_LEVEL EQUAL 1 )
-   set( AUDACITY_SUFFIX "-beta" )
-else()
-   set( AUDACITY_SUFFIX "" )
+set( AUDACITY_REVISION 0 )
+find_package( Git QUIET )
+if( GIT_FOUND )
+    execute_process(
+        COMMAND
+            ${GIT_EXECUTABLE} describe --abbrev=0
```
Comment (Be-ing) [inline CMakeLists.txt]: Why did you change this? This only shows the last tag now.

Comment (Be-ing) [reply CMakeLists.txt]: Oh I see Git is called again below... that's confusing.

Comment (nbsp) [reply CMakeLists.txt]: That's also how it was before the latest revision. I could move the first one down and couple both together, now that the "version" doesn't matter as much.

File: CMakeLists.txt [hunk 5]
```diff
@@ -232,32 +216,44 @@ elseif( CMAKE_SYSTEM_NAME MATCHES "Darwin" )
...
-      list( GET git_output 1 GIT_COMMIT_LONG )
-      list( GET git_output 2 GIT_COMMIT_TIME )
-   endif()
+    execute_process(
+        COMMAND
+            ${GIT_EXECUTABLE} describe --abbrev=7
+        WORKING_DIRECTORY
+            ${CMAKE_SOURCE_DIR}
+        OUTPUT_VARIABLE
+            git_output
```
Comment (Be-ing) [inline CMakeLists.txt]:
```suggestion
            GIT_DESCRIBE
```

File: CMakeLists.txt [hunk 6]
```diff
@@ -232,32 +216,44 @@ elseif( CMAKE_SYSTEM_NAME MATCHES "Darwin" )
...
+        OUTPUT_STRIP_TRAILING_WHITESPACE
+        ERROR_QUIET
+    )
+
+    if( git_output )
+        set( GIT_DESCRIBE ${git_output} )
```
Comment (Be-ing) [inline CMakeLists.txt]:
```suggestion
    if( GIT_DESCRIBE )
        # copy to new variable for string manipulation
        set( git_output ${GIT_DESCRIBE} )
```

File: CMakeLists.txt [hunk 7]
```diff
@@ -232,32 +216,45 @@ elseif( CMAKE_SYSTEM_NAME MATCHES "Darwin" )
...
+            ${GIT_EXECUTABLE} describe --abbrev=7
+        WORKING_DIRECTORY
+            ${CMAKE_SOURCE_DIR}
+        OUTPUT_VARIABLE
+            GIT_DESCRIBE
+        OUTPUT_STRIP_TRAILING_WHITESPACE
+        ERROR_QUIET
+    )
+
+    if( git_output )
```
Comment (Be-ing) [inline CMakeLists.txt]:
```suggestion
    if( GIT_DESCRIBE )
```

Comment (nbsp) [reply CMakeLists.txt]: Whoops.

**Review by Be-ing — APPROVED** [review]

**Review by Semisol — COMMENTED** [review]
uhm something is breaking
`clang: error: argument to '-Xarch_x86_64' is missing (expected 1 value)`

**Comment by nbsp** [comment]
> uhm something is breaking
> `clang: error: argument to '-Xarch_x86_64' is missing (expected 1 value)`

For the record: it was a problem with sccache.

**Review by Semisol — APPROVED** [review]
switched to ccache and it worked
I'll let you merge it

**Comment by nbsp** [comment]
I'll merge it when sourcehut comes back up. https://status.sr.ht
